lib/Target/X86/X86FrameLowering.cpp

```diff
@@ -393,28 +393,152 @@ static bool usesTheStack(const MachineFunction &MF) {
-  // as future work. 
-  BuildMI(MBB, MBBI, DL, TII.get(X86::SUB64rr), X86::RSP)
-    .addReg(X86::RSP)
+  // MBB:
+  //    ZeroReg = 0
+  //    CopyReg = RSP
+  //    Flags, TestReg = CopyReg - RAX
+  //    FinalReg = !Flags.Ovf ? TestReg : ZeroReg
+  //    LimitReg = gs magic thread env access
+  //    if FinalReg >= LimitReg goto ContinueMBB
+  // RoundBB:
+  //    RoundReg = page address of FinalReg
+  // LoopMBB:
+  //    LoopReg = PHI(LimitReg,ProbeReg)
+  //    ProbeReg = LoopReg - PageSize
+  //    [ProbeReg] = 0
+  //    if (ProbeReg > RoundReg) goto LoopMBB
+  // ContinueMBB:
+  //    RSP = RSP - RAX
+  //    [rest of original MBB]
+
+  // Set up the new basic blocks
+  MachineBasicBlock *RoundMBB = MF.CreateMachineBasicBlock(LLVM_BB);
+  MachineBasicBlock *LoopMBB = MF.CreateMachineBasicBlock(LLVM_BB);
+  MachineBasicBlock *ContinueMBB = MF.CreateMachineBasicBlock(LLVM_BB);
+
+  MachineFunction::iterator MBBIter = MBB;
+  ++MBBIter;
+
+  MF.insert(MBBIter, RoundMBB);
+  MF.insert(MBBIter, LoopMBB);
+  MF.insert(MBBIter, ContinueMBB);
+
+  // Split MBB and move the tail portion down to ContinueMBB.
+  ContinueMBB->splice(ContinueMBB->begin(), &MBB, std::next(MBBI), MBB.end());
+  ContinueMBB->transferSuccessorsAndUpdatePHIs(&MBB);
+
+  // Some useful constants
+  const int64_t ThreadEnvironmentStackLimit = 0x10;
+  const int64_t PageSize = 0x1000;
+  const int64_t PageMask = ~(PageSize - 1);
+
+  // Virtual registers we need.
+  MachineRegisterInfo &MRI = MF.getRegInfo();
+  const TargetRegisterClass *RegClass = &X86::GR64RegClass;
+  const unsigned ZeroReg = MRI.createVirtualRegister(RegClass),
+    CopyReg = MRI.createVirtualRegister(RegClass),
+    TestReg = MRI.createVirtualRegister(RegClass),
+    FinalReg = MRI.createVirtualRegister(RegClass),
+    RoundedReg = MRI.createVirtualRegister(RegClass),
+    LimitReg = MRI.createVirtualRegister(RegClass),
+    JoinReg = MRI.createVirtualRegister(RegClass),
+    ProbeReg = MRI.createVirtualRegister(RegClass);
+
+  // Add code to MBB to check for overflow and set the new target stack pointer
+  // to zero if so.
+  BuildMI(&MBB, DL, TII.get(X86::XOR64rr), ZeroReg)
+    .addReg(ZeroReg, RegState::Undef)
+    .addReg(ZeroReg, RegState::Undef);
+  BuildMI(&MBB, DL, TII.get(X86::MOV64rr), CopyReg).addReg(X86::RSP);
+  BuildMI(&MBB, DL, TII.get(X86::SUB64rr), TestReg)
+    .addReg(CopyReg)
     .addReg(X86::RAX);
+  BuildMI(&MBB, DL, TII.get(X86::CMOVB64rr), FinalReg)
+    .addReg(TestReg)
+    .addReg(ZeroReg);
+
+  // FinalReg now holds final stack pointer value, or zero if allocation would
+  // overflow. Compare against the current stack limit from the thread
+  // environment block.
+  BuildMI(&MBB, DL, TII.get(X86::MOV64rm), LimitReg)
+    .addReg(0)
+    .addImm(1)
+    .addReg(0)
+    .addImm(ThreadEnvironmentStackLimit)
+    .addReg(X86::GS);
+  BuildMI(&MBB, DL, TII.get(X86::CMP64rr)).addReg(FinalReg).addReg(LimitReg);
```

Comment:
I doubt it would ever get folded since LimitReg is live beyond this point so needs to be in a register. And at least in current usage it is not folded:
```
00007FFA65A40E59 4D 0F 42 D3          cmovb       r10,r11  
00007FFA65A40E5D 65 4C 8B 1C 25 10 00 00 00 mov         r11,qword ptr gs:[10h]  
00007FFA65A40E66 4D 39 DA             cmp         r10,r11  
```